xlml/apis/task.py

```diff
@@ -70,14 +70,91 @@ def run(self) -> DAGNode:
         group_id=self.task_test_config.benchmark_id, prefix_group_id=True
     ) as group:
       provision, queued_resource, ssh_keys = self.provision()
-      run_model = self.run_model(queued_resource, ssh_keys)
       post_process = self.post_process()
       clean_up = self.clean_up(queued_resource)
 
-      provision >> run_model >> post_process >> clean_up
+      if self.task_test_config.startup_script:
+        check_if_startup_script_end = self.check_if_startup_script_end(
+            queued_resource, ssh_keys
+        )
+        provision >> check_if_startup_script_end >> post_process >> clean_up
```

Comment:
I am a little bit confused about this startup_script logic. I think it should be inside of `provision` group when requesting the resource ([link](https://github.com/GoogleCloudPlatform/ml-auto-solutions/blob/master/xlml/utils/tpu.py#L75)). Can you explain a little bit more? Thanks!

Reply:
Startup script will be run once the QR is active: https://cloud.google.com/tpu/docs/managing-tpus-tpu-vm#creating_a_cloud_tpu_queued_resource_using_gcloud

The order would be:
Create QR with startup script -> Wait QR to be active/ready -> Startup script will be run. 

Two things:
1. Startup script will be run once QR is ready, so I can't check the workload status until `wait_for_ready_queued_resource` is done.
2. Once the startup script is running, I couldn't tell when will the workload end. So I have function `check_if_startup_script_end` to ssh into 1 worker to check if the workload finished by the following 2 files:
  * `/tmp/main_process_id.txt` has the process id of the main workload
  * `/tmp/process_exit_status.txt` has the exit status of the main workload
  * `/tmp/logs` has the logs of the main workload

Reply:
Yes, I have the same thought that I want to put everything into run().

My startup script work flow is `provision >> check_if_startup_script_end >> post_process >> clean_up`, which `provision` part contain `startup script`, `check_if_startup_script_end` will check it workload finished and print out logs.

Reply:
Thanks Tony for the explanation! In the workflow `provision >> check_if_startup_script_end >> post_process >> clean_up`, customers cannot use both startup_script feature and run_model feature. I understand we don't have a solid use case yet, but it is possible in the future. 

`Once the startup script is running, I couldn't tell when will the workload end`. I see, in this case, I think it's better we have the workflow like (adding `wait_for_startup_script_completion`): 

* Provision
  * initialize (generate_tpu_name & generate_ssh_keys)
  * create_queued_resource (create_queued_resource_request, wait_for_ready_queued_resource & wait_for_startup_script_completion)  
  * setup (keep this step: 1) run it when there is script; 2) run [empty operator](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/empty/index.html) when no script is passed )
* run_model (similar to setup)
* post_process (keep the same)
* clean_up (keep the same)

You will also need to handle the job_status properly based on those changes: https://github.com/GoogleCloudPlatform/ml-auto-solutions/blob/master/xlml/utils/metric.py#L503

Reply:
Just restructured it as you suggested.

with startup script:
provision (CQR+wait+check_if_startup_script_end) -> run_model(dummy dag that do nothing) -> post_process -> clean_up
https://screenshot.googleplex.com/3QGybCApV8Sux59

Without startup script:
provision (CQR+wait) -> setup -> run_model -> post_process -> clean_up
https://screenshot.googleplex.com/8rC7VRALGtHvEJd